engines/docker/engine.go

```diff
@@ -0,0 +1,97 @@
+package dockerengine
+
+import (
+	"sync"
+
+	docker "github.com/fsouza/go-dockerclient"
+	schematypes "github.com/taskcluster/go-schematypes"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/caching"
+)
+
+type engine struct {
+	engines.EngineBase
+	m           sync.Mutex
+	Environment *runtime.Environment
+	client      *docker.Client
+	monitor     runtime.Monitor
+	config      configType
+	running     int
+	cache       *caching.Cache
+}
+
+type engineProvider struct {
+	engines.EngineProviderBase
+}
+
+func init() {
+	engines.Register("docker", engineProvider{})
+}
+
+func (p engineProvider) ConfigSchema() schematypes.Schema {
+	return configSchema
+}
+
+func (p engineProvider) NewEngine(options engines.EngineOptions) (engines.Engine, error) {
+	debug("docker engineProvider.NewEngine()")
+	var c configType
+	schematypes.MustValidateAndMap(configSchema, options.Config, &c)
+	if c.DockerSocket == "" {
+		c.DockerSocket = "unix:///var/run/docker.sock"
+	}
+
+	client, err := docker.NewClient(c.DockerSocket)
+	if err != nil {
+		return nil, err
```

Comment:
`return nil, errors.Wrapf(err, "failed to connect to dockerSocket: '%s'", c.DockerSocket)`

The [`"github.com/pkg/errors"`](https://godoc.org/github.com/pkg/errors) package can be used as drop-in replacement for the built-in `"errors"` package. And it has `errors.Wrap(err, message)` and `errors.Wrapf(err, fmt, arg...)`.

If ever passing up an unhandled error like this, wrapping it helps a lot... it makes it's very clear what when wrong.
Sometimes libraries just return a generic OS error.. like "file not found", which is hard to debug... If we wrap it with `failed to connect to dockerSocket: 'unix:///var/run/docker.sock'` it becomes very obvious, that we failed to connect to docker-socket because the file was missing :)

Note: we could also use `monitor.ReportError` and then return `runtime.ErrFatalInternalError`, but returning a wrapped error is also perfectly fine in this case... As the caller will take care to note that the error happened in `NewEngine` before logging it and reporting it. Also note that if you just return `runtime.ErrFatalInternalError` then caller will assume you've already reported it to sentry and printed it to error log with `monitor.Error(...)`